### ISSUE TITLE
Can't import recepies

### ISSUE THREAD

**Issue by JoshuaPettus** [nextcloud/cookbook#722]
Not really a bug report, as this was working for me and now I can't import any recepies.  I don't see any tickets on this open so It's most likely something got screwy with my setup.

When I import any webpage, I just get a white dialog box that pops up with Ok.  No error warning.

This recipe for instance: https://www.tastesoflizzyt.com/homemade-apple-butter/

In the log I get a bunch of errors like this

```
{
  "reqId": "yYhy5hs54OWWcKMaaIVP",
  "level": 3,
  "time": "2021-05-17T20:52:23+00:00",
  "remoteAddr": "---",
  "user": "----",
  "app": "index",
  "method": "POST",
  "url": "/apps/cookbook/import",
  "message": {
    "Exception": "Exception",
    "Message": "preg_match(): Argument #2 ($subject) must be of type string, array given",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/nextcloud/lib/private/AppFramework/App.php",
        "line": 157,
        "function": "dispatch",
        "class": "OC\\AppFramework\\Http\\Dispatcher",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/lib/private/Route/Router.php",
        "line": 302,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::"
      },
      {
        "file": "/var/www/nextcloud/lib/base.php",
        "line": 993,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/index.php",
        "line": 37,
        "function": "handleRequest",
        "class": "OC",
        "type": "::"
      }
    ],
    "File": "/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
    "Line": 159,
    "Previous": {
      "Exception": "TypeError",
      "Message": "preg_match(): Argument #2 ($subject) must be of type string, array given",
      "Code": 0,
      "Trace": [
        {
          "file": "/var/www/nextcloud/apps/cookbook/lib/Service/RecipeService.php",
          "line": 218,
          "function": "preg_match"
        },
        {
          "file": "/var/www/nextcloud/apps/cookbook/lib/Service/RecipeService.php",
          "line": 488,
          "function": "checkRecipe",
          "class": "OCA\\Cookbook\\Service\\RecipeService",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/apps/cookbook/lib/Service/RecipeService.php",
          "line": 829,
          "function": "parseRecipeHtml",
          "class": "OCA\\Cookbook\\Service\\RecipeService",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/apps/cookbook/lib/Controller/MainController.php",
          "line": 351,
          "function": "downloadRecipe",
          "class": "OCA\\Cookbook\\Service\\RecipeService",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 218,
          "function": "import",
          "class": "OCA\\Cookbook\\Controller\\MainController",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php",
          "line": 127,
          "function": "executeController",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/lib/private/AppFramework/App.php",
          "line": 157,
          "function": "dispatch",
          "class": "OC\\AppFramework\\Http\\Dispatcher",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/lib/private/Route/Router.php",
          "line": 302,
          "function": "main",
          "class": "OC\\AppFramework\\App",
          "type": "::"
        },
        {
          "file": "/var/www/nextcloud/lib/base.php",
          "line": 993,
          "function": "match",
          "class": "OC\\Route\\Router",
          "type": "->"
        },
        {
          "file": "/var/www/nextcloud/index.php",
          "line": 37,
          "function": "handleRequest",
          "class": "OC",
          "type": "::"
        }
      ],
      "File": "/var/www/nextcloud/apps/cookbook/lib/Service/RecipeService.php",
      "Line": 218
    },
    "CustomMessage": "--"
  },
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0",
  "version": "21.0.1.1",
  "id": "60a2d90383a2f"
}
{
  "reqId": "yYhy5hs54OWWcKMaaIVP",
  "level": 2,
  "time": "2021-05-17T20:52:23+00:00",
  "remoteAddr": "---",
  "user": "---",
  "app": "cookbook",
  "method": "POST",
  "url": "/apps/cookbook/import",
  "message": "libxml: Error 76 occurred 1 times while parsing https://www.tastesoflizzyt.com/homemade-apple-butter/. Last time in line 4434 and column 2068: Unexpected end tag : div\n",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0",
  "version": "21.0.1.1",
  "id": "60a2d90383a4a"
}
{
  "reqId": "yYhy5hs54OWWcKMaaIVP",
  "level": 2,
  "time": "2021-05-17T20:52:23+00:00",
  "remoteAddr": "---",
  "user": "---",
  "app": "cookbook",
  "method": "POST",
  "url": "/apps/cookbook/import",
  "message": "libxml: Error 513 occurred 28 times while parsing https://www.tastesoflizzyt.com/homemade-apple-butter/. Last time in line 4403 and column 17: ID twitter already defined\n",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0",
  "version": "21.0.1.1",
  "id": "60a2d90383a62"
}
```

Any Ideas?

**Comment by JoshuaPettus** [comment]
I should have noted I'm using NC 21 with PHP 8.0.5

**Comment by christianlupus** [comment]
I just checked and I found the recipe to be importable. Some errors in the logs are there as the page is not 100% HTML conformant. The internal XML parser in PHP is quite strict throwing dozens of warnings.

The other problem is in [this line](https://github.com/nextcloud/cookbook/blob/e935b2f0e9c31989ccce926aca44daf7e2f5d2e5/lib/Service/RecipeService.php#L219). The recipe as published currently has the following property:
```
...
"recipeYield": ["20"],
...
```
This cannot be used with a regular match as the regex assumes a string automatically.

I can merge a quick fix but this should be covered in the future more in depth.
